ENH: Simplify and speed up navigation bar links generation

choldgraf opened the pull request:
This simplifies our header navigation link code, so that we manually generate these links instead of using the Sphinx `TocTree` function. It does this by creating a new function to use in our HTML templates. Because we call this *twice* for the header, this should make documentation builds much faster.

The output shouldn't be much different, though it will remove some Sphinx-specific classes from the output nav items in the header (like `toctree-l1`). It also means that "numbered" top-level sections don't have numbers anymore (but the sidebars will have numbers like normal). I am not sure how to re-implement that manually but it seems like a reasonable trade-off to me if we get a big speed boost.

cc @tupui and maybe @drammock who I think both mentioned page builds had gone up.

If this makes things faster, I think it:

- closes https://github.com/pydata/pydata-sphinx-theme/issues/876
- closes https://github.com/pydata/pydata-sphinx-theme/issues/859
- closes https://github.com/pydata/pydata-sphinx-theme/issues/855

tupui commented on tests/sites/sidebars/_templates_single_sidebar/components/sidebar-nav-bs.html:
```diff
@@ -1,6 +1,6 @@
   <div class="bd-toc-item active">
     <!-- Specify a startdepth of 0 instead of default of 1 -->
-    {{ generate_nav_html("sidebar", startdepth=0, maxdepth=4, collapse=True, includehidden=True, titles_only=True) }}
+    {{ generate_toctree_html("sidebar", startdepth=0, maxdepth=4, collapse=True, includehidden=True, titles_only=True) }}
```
FYI this will be a breaking change at least for SciPy since we are using this in a template (`sidebar-nav-bs.html`). Putting the template here in case we do something really wrong.

```
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <div class="bd-toc-item active">
    {% if pagename.startswith("reference") %}
    {{ generate_nav_html("sidebar", maxdepth=1, collapse=True, includehidden=True, titles_only=True) }}
    {% else %}
    {{ generate_nav_html("sidebar", maxdepth=2, collapse=True, includehidden=True, titles_only=True) }}
    {% endif %}
  </div>
</nav>
```

choldgraf replied:
ah, we can change this back if need be, I just thought that `toctree_html` was a more descriptive name. Maybe we can just add a deprecation warning

tupui replied:
As I either way need to adapt more things I am personally fine with this change. But yes if this was supposed to be public, having a warning would be nice.

tupui replied:
By the way, can this template be the cause of this formatting issue I now see:
<img width="390" alt="Screenshot 2022-08-17 at 16 40 03" src="https://user-images.githubusercontent.com/23188539/185166979-d3655d0d-0853-42b0-9e08-27274550d19f.png">

On 0.9 it's fine but not on the RC or here.

choldgraf replied:
there's now a deprecation warning for this function, and I can't see this checkbox issue on the scipy docs, so I'll resolve this and can you open a new issue with a reproducible example of this issue on the scipy docs if it's still there? (w/ the extra checkbox)

tupui reviewed (commented):
Sorry still not working for SciPy...

python dev.py doc -j 6  4025.48s user 221.28s system 434% cpu 16:18.01 total

vs 

python dev.py doc -j 6  743.65s user 61.77s system 438% cpu 3:03.52 total

Let me know if I can help with some timing or other (I don't know how and what so I would need directions, we can also make a call if you want.)

choldgraf commented:
can you provide more context what your commands are? They look to be the exact same command with different results

tupui commented:
Sorry yes, first is building with this PR and the other is building with 0.9. The rest is identical with Sphinx 5.1.1.

choldgraf commented:
Gotcha - well then I am really not sure what's causing this slow-down. In my opinion, this PR still improves things though

tupui commented:
At least on my side I didn't notice a difference with the current RC2. I have

python dev.py doc -j 6  4204.69s user 179.17s system 462% cpu 15:48.39 total

choldgraf commented:
I mean improves things in terms of simplicity rather than speed, but if others disagree I can just close this PR.

choldgraf commented:
maybe @jorisvandenbossche can try profiling the arrow etc docs with this version as well?

tupui commented:
@drammock I've seen you pushed some changes. Let me know if I should try to build again. Or again if there is some other diagnostic I can run to help (I might need guidance here).

drammock commented:
OK I tried building MNE-Python docs to check for slowdowns

#### on pydata-sphinx-theme 5d0c6c2c (this PR)
```
231.16user 2.51system 3:52.78elapsed 100%CPU (0avgtext+0avgdata 1010004maxresident)k
427384inputs+522559outputs (884major+316128minor)pagefaults 0swaps
```

#### on pydata-sphinx-theme 83e39b97 (current main)
```
256.93user 2.61system 4:18.82elapsed 100%CPU (0avgtext+0avgdata 1015004maxresident)k
468098inputs+521394outputs (0major+316763minor)pagefaults 0swaps
```

#### on pydata-sphinx-theme b97d3321 (v0.9.0)
```
257.40user 2.77system 4:19.37elapsed 100%CPU (0avgtext+0avgdata 1020516maxresident)k
442204inputs+515520outputs (0major+319835minor)pagefaults 0swaps
```

#### environment

<details>

```
$ mamba list sphinx
# packages in environment at /opt/miniforge3/envs/mnedev:
#
# Name                    Version                   Build  Channel
jupyter-sphinx            0.4.0                    pypi_0    pypi
jupyter_sphinx            0.4.0           py310hff52083_0    conda-forge
nbsphinx                  0.8.9              pyhd8ed1ab_0    conda-forge
pydata-sphinx-theme       0.10.0rc2                pypi_0    pypi
sphinx                    5.1.1              pyhd8ed1ab_1    conda-forge
sphinx-autobuild          2021.3.14          pyhd8ed1ab_0    conda-forge
sphinx-copybutton         0.5.0                    pypi_0    pypi
sphinx-design             0.2.0              pyhd8ed1ab_1    conda-forge
sphinx-gallery            0.11.0.dev0               dev_0    <develop>
sphinx-sitemap            2.2.0              pyhd8ed1ab_0    conda-forge
sphinx-theme-builder      0.2.0b1            pyhd8ed1ab_0    conda-forge
sphinx-togglebutton       0.3.2                    pypi_0    pypi
sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
sphinxcontrib-bibtex      2.4.2              pyhd8ed1ab_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    2.0.0              pyhd8ed1ab_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.5              pyhd8ed1ab_2    conda-forge
sphinxcontrib-youtube     1.2.0                    pypi_0    pypi
sphinxext-rediraffe       0.2.7              pyhd8ed1ab_0    conda-forge
```

</details>

so for me, things are in fact faster on this PR than on `main` or `v0.9.0`

I'll see what I can do WRT reproducing the scipy build, but I've had difficulty in the past getting scipy dev environment set up on my machine, so if that continues I may need to rely on @tupui to help debug

tupui commented:
@drammock nice I can try to build with that now.

I am more than happy to help you setup SciPy if you want. We can make a quick call on discord or anything else if that helps (feel free to send me an invite too).

TL;DR we have updated our quick start instructions to build: https://scipy.github.io/devdocs/dev/dev_quickstart.html

tupui commented:
Just tested again and I still have slow builds I am afraid.

drammock commented:
@tupui OK I think I have a working scipy dev env under the new meson build system. `python dev.py test` yields this:
```
FAILED scipy/linalg/tests/test_solvers.py::test_solve_discrete_are - AssertionError: 
= 1 failed, 37153 passed, 2172 skipped, 12121 deselected, 133 xfailed, 13 xpassed in 374.61s (0:06:14) ==
```

Docs say all test should pass on Linux but I'm seeing one failure, will assume it's a fluke and not something wrong with my install / dependencies.  LMK if that's a failure you've seen before and know how to fix it.  Meanwhile I think I can now at least try to replicate your slow builds.

tupui commented:
Super thanks! It's fine for the tests especially if it's just one.

For the doc you can build using:

`python dev.py doc -j N_PROCS` with N_PROCS the number of cores you want to use to have it parallel. Or you do it manually with make going to the doc folder and it should work as well.

tupui commented:
You can use my branch here which has updates for the new version https://github.com/scipy/scipy/pull/16660

drammock commented:
Hmm, after removing v0.9 of the theme from the conda env and reinstalling it with `pip install -e .`, I can't even get the scipy docs to build:

```
Traceback (most recent call last):
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/sphinx/builders/html/__init__.py", line 1048, in handle_page
    output = self.templates.render(templatename, ctx)
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/sphinx/jinja2glue.py", line 188, in render
    return self.environment.get_template(template).render(context)
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/sphinx/themes/basic/page.html", line 10, in top-level template code
    {%- extends "layout.html" %}
  File "/opt/scipy/doc/source/_templates/layout.html", line 12, in top-level template code
    {% extends "!layout.html" %}
  File "/opt/pydata-sphinx-theme/src/pydata_sphinx_theme/theme/pydata_sphinx_theme/layout.html", line 25, in top-level template code
    {% set remove_sidebar_secondary = (meta is defined and meta is not none
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/sphinx/themes/basic/../basic/layout.html", line 169, in top-level template code
    {%- block content %}
  File "/opt/pydata-sphinx-theme/src/pydata_sphinx_theme/theme/pydata_sphinx_theme/layout.html", line 85, in block 'content'
    {% include "sections/sidebar-primary.html" %}
  File "/opt/pydata-sphinx-theme/src/pydata_sphinx_theme/theme/pydata_sphinx_theme/sections/sidebar-primary.html", line 1, in top-level template code
    {% block docs_sidebar %}
  File "/opt/pydata-sphinx-theme/src/pydata_sphinx_theme/theme/pydata_sphinx_theme/sections/sidebar-primary.html", line 34, in block 'docs_sidebar'
    {%- include sidebartemplate %}
  File "/opt/scipy/doc/source/_templates/sidebar-nav-bs.html", line 6, in top-level template code
    {{ generate_nav_html("sidebar", maxdepth=2, collapse=True, includehidden=True, titles_only=True) }}
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/jinja2/sandbox.py", line 391, in call
    if not __self.is_safe_callable(__obj):
  File "/opt/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/jinja2/sandbox.py", line 275, in is_safe_callable
    getattr(obj, "unsafe_callable", False) or getattr(obj, "alters_data", False)
jinja2.exceptions.UndefinedError: 'generate_nav_html' is undefined
```

It looks like the error is in your template of `sidebar-nav-bs`: 

```
  File "/opt/scipy/doc/source/_templates/sidebar-nav-bs.html", line 6, in top-level template code
    {{ generate_nav_html("sidebar", maxdepth=2, collapse=True, includehidden=True, titles_only=True) }}
```
it's calling `generate_nav_html` and not finding it.  Are you not seeing that?

drammock commented:
ah OK, I see, I should be using your open PR branch from https://github.com/scipy/scipy/pull/16660

drammock commented:
running with `-j 12`, here are the build times I get for scipy

#### on pydata-sphinx-theme 5d0c6c2 (this PR)

```
5913.52user 147.10system 9:49.39elapsed 1028%CPU (0avgtext+0avgdata 1794204maxresident)k
66180587inputs+1258600outputs (4067major+17010014minor)pagefaults 0swaps
```

#### on pydata-sphinx-theme 83e39b97 (main)
Cannot build docs, `UndefinedError("'generate_toctree_html' is undefined")`

#### on pydata-sphinx-theme  (v0.9.0)
Didn't try because assuming same error as on `main`

@tupui I need more details about how exactly you are testing / timing.  Given what is going on in scipy's `doc/source/_templates/sidebar-nav-bs.html` I can't seem to actually do a fair comparison (?)

tupui commented:
@drammock To build on main, you just need to remove my last commit. This is just a change in the template name introduced in this PR (on the theme) that I needed to fix in my PR.

choldgraf commented:
If it simplifies things to undo the function renaming, that is fine with me. I just renamed it because I thought it would make things easier to remember and understand.

It is a bit strange that @drammock is reporting a boost in performance while @tupui is not 🤔

drammock commented:
actually, just thinking about it, it appears that scipy does not override the theme's `layout.html` (which means the theme's `layout.html` file will call `generate_toctree_html`), but in scipy's `doc/source/_templates/sidebar-nav-bs.html` that function is being called *again*.  If that's correct, I think in effect you are manually doing twice what we were previously doing twice automatically (and which this PR fixed to only run once by default).  Can you investigate/confirm?

tupui commented:
Oh then you mean that this should just be moved to `layout.html`?

drammock commented:
> Oh then you mean that this should just be moved to `layout.html`?

well, I haven't looked in detail about exactly what you're doing in the `sidebar-nav-bs` template.  I just followed the traceback and looked at a couple lines of context.  Let me look...

@choldgraf It looks like what scipy is doing is adjusting `maxdepth` depending on which page is being rendered (1 for "reference" pages and 2 for other pages).  You understand sphinx better than I do, is there a way to achieve that without having to build the toctree twice?  Can the toctree be built once, and then simply "pruned" before being rendered on certain pages?

@tupui is it really necessary to limit the maxdepth in this way, or would it work to instead control the level of "collapse" (https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/navigation.html?highlight=collapse#control-how-many-navigation-levels-are-shown-by-default) ?

tupui commented:
We did this because the tree is huge for SciPy so for all API we just want a single level otherwise the HTML is huge just for the tree (and since we have thousands of functions, then the overall size of the doc explode just for that). For the other pages we wanted to have a second level, but if that's too complicated then a single level could do. Ideally, this tree should not be repeated on all HTML pages, but I understood from previous discussions that it was not possible or difficult to do.

Right now I am trying to set `"navigation_depth": 1` and remove the template, to see if it improves things.

choldgraf commented:
Maybe this is something we could also add a feature to control in a per page basis?

drammock commented:
OK now I've got all the scipy timings (running with `-j 12`).  There is definitely a big difference.

#### on pydata-sphinx-theme 5d0c6c2 (this PR)

```
5913.52user 147.10system 9:49.39elapsed 1028%CPU (0avgtext+0avgdata 1794204maxresident)k
66180587inputs+1258600outputs (4067major+17010014minor)pagefaults 0swaps
```

#### on pydata-sphinx-theme 83e39b97 (main)
```
6263.57user 146.89system 10:16.43elapsed 1039%CPU (0avgtext+0avgdata 1801600maxresident)k
66306646inputs+1254338outputs (4067major+17158334minor)pagefaults 0swaps
```

#### on pydata-sphinx-theme b97d3321 (v0.9.0)
```
1046.83user 133.36system 2:05.87elapsed 937%CPU (0avgtext+0avgdata 1803232maxresident)k
66240519inputs+1245583outputs (4067major+11678283minor)pagefaults 0swaps
```

tupui commented:
I have the timing on my side for when I set `"navigation_depth": 1` and remove the template

```
python dev.py doc -j 6  1775.51s user 144.60s system 396% cpu 8:04.80 total
```

Which is way better but still far from on 0.9

```
python dev.py doc -j 6 743.65s user 61.77s system 438% cpu 3:03.52 total
```

drammock commented:
OK the culprit is #744 

#### on pydata-sphinx-theme 9cc2486978e044ed0273c207a9c8bb0308a5e3aa (just before #744) and scipy e792b3acb:
```
1036.01user 130.35system 2:06.07elapsed 925%CPU (0avgtext+0avgdata 1799460maxresident)k
66261672inputs+1246308outputs (4067major+11584651minor)pagefaults 0swaps
```

#### on pydata-sphinx-theme bbc855b6d7bf402c9b5195d832128e564a4e6cb6 (after #744) and scipy e792b3acb:
```
6128.59user 145.32system 10:02.05elapsed 1042%CPU (0avgtext+0avgdata 1801480maxresident)k
66289941inputs+1243865outputs (4067major+16914748minor)pagefaults 0swaps
```

choldgraf commented:
OK I looking at https://github.com/pydata/pydata-sphinx-theme/pull/744, I suspect that the problematic lines are this:

https://github.com/pydata/pydata-sphinx-theme/pull/744/files#diff-48a05e3e541868308449c2edfb57f539962c5748f212d7282b6257ba0003e9adR2-R9

So I decided to try a simple fix to see if this works. I've added an `lru_cache` to the `generate_toctree_html` function. Theoretically this means if it's being called twice w/ the same arguments, it should just quickly return the result. Maybe that will improve things? Can @tupui or @drammock check?

I also added a deprecation for the old `generate_nav_html` function.

tupui commented:
Thank you both! I will see if I have time to test this WE otherwise I do that first thing on Monday.

tupui commented:
I have a bit better (still `"navigation_depth": 1` and remove the template)

```
python dev.py doc -j 6  1767.14s user 138.66s system 429% cpu 7:23.92 total
```

choldgraf commented:
That seems like a pretty big improvement to me!

Would it be helpful if you could specify the navigation depth on a per-page basis in the metadata? like:

```
:html_theme.navigation_depth: 1
```

tupui commented:
Yes it's a nice improvement compared to the previous commit. What I meant is that it's still x2 slower than the current release.

> Would it be helpful if you could specify the navigation depth on a per-page basis in the metadata?

It would although sounds more complicated than editing the template. But if there is no way around sure I will adapt 😃

Circling back on that, there is really no way to have the whole TOC not included in every html pages? Like having some JS that would call an existing resource?

choldgraf commented:
There probably is some way to reuse an existing resource but I think it would make the theme a lot more complex to maintain because it would basically circumvent how sphinx does navigation.

drammock commented:
> there is really no way to have the whole TOC not included in every html pages?

Just thinking it loud here: It might be possible by overriding the theme's templates so that the TOC is only built once (in an orphan page perhaps) and all other pages use a sidebar template containing the relevant code to dynamically put that orphan TOC in the page at load time. This would be tricky though to make things like scrollspy and expand/collapse happen correctly, you might need some JS for that too. All this to say I don't see (yet?) why it would be impossible, but it's complicated enough that I doubt we'll build it into the theme, and it may in fact be impossible for reasons I haven't thought of.

choldgraf commented:
What is the minimal speed increase we need to hit in order to consider this PR worth implementing? I don't have much time remaining to work on this, so I think it's a question of "is this enough of an iterative improvement, are there _minimal_ changes to make here, or should we just close this PR?".

tupui commented:
I will try to use this PR in SciPy's CI. On our side at least, this would be the determinant factor. If we can built at all or not.

tupui commented:
Ok the doc managed to build without timeout. So in our CI we go from 20 (0.9) to 45 mins (this PR). In detail, without the artifact upload there is roughly a factor 3 (12 vs 36 mins).

choldgraf commented:
I've added a GitHub action to this repository that will generate a `py-spy` results SVG and upload it as an artifact in a GitHub action. If somebody wants to use that to compare the call stack between PRs, we can use that now.

For example:

- [Here's the artifact for this PR](https://github.com/pydata/pydata-sphinx-theme/suites/7921564015/artifacts/337526118)
- [Here's the artifact on main](https://github.com/pydata/pydata-sphinx-theme/suites/7921546745/artifacts/337528592)

I'm way over my "nights and weekends" budget on this theme so I will be able to make minimal improvements on this one. However I am more than happy to review PRs or suggestions from others for how we can improve the performance.

tupui commented:
To be clear, I am not blaming you for anything and really appreciate all the work you all do on the theme 😃 I am also spending ludicrous time on OSS (not counting my offset from work).

choldgraf commented:
I looked into this a little bit more, and something tells me that this is an issue with SciPy's implementation, and not necessarily going to be experienced by others.

I used the `timeit.repeat` function to time the builds of the test site when adding 25 extra pages to increase its length a bit.

Here are the results with the **current PR**:

```
Time per build: [1.3600825106647487, 1.3852431703344337, 1.4198622083325365]
```

Here are the results with **v0.9**:

```
Time per build: [1.4451562219983316, 1.645955220000663, 1.7013943913334515]
```

So at least with this test, this PR is actually _faster_ than `v0.9`.

tupui commented:
Thank you for doing this! I think you are right @choldgraf. I tried to slim down the doc and just compile one or two submodules and I have similar timings. I suppose I must then try to add them one by one until I see diverging timings... I don't know how to debug this better.

drammock commented:
you could also try removing scipy's custom `_template` overrides, then adding back elements of the templates one at a time.

tupui commented:
Aside from the `autosummary` part, the last version I am testing is without anything else (as above).

choldgraf commented:
What do people think we should do with this PR? According to the tests in https://github.com/pydata/pydata-sphinx-theme/pull/878#issuecomment-1223737760, this PR is a significant performance improvement over `0.9`. However, it seems like https://github.com/pydata/pydata-sphinx-theme/issues/855 will still not be resolved, but may also no be related to the pydata theme directly?

Can we merge this in as an iterative improvement (happy to have a review of the code though) and figure out what to do next in follow-up issues?

tupui reviewed (commented):
I am personally fine with iterating elsewhere if you want to get this one in. It does bring an improvement over `main`.

tupui commented:
I did a few experiments and I am really puzzled. Individual modules seems to compile at a similar speed between the versions. But when grouped together, some slowdown appears. e.g. the sparse module is being built fast when it's alone, but when there are other modules it seems like it's building slower for some reason.

Maybe there are some deadlocks somewhere. I am not sure what I can try next.

tupui commented:
Oh @drammock I actually just tested to remove all autosummary templates and this is it! But seems like it's really a problem only when building multiple modules.

drammock commented:
> I actually just tested to remove all autosummary templates and this is it! But seems like it's really a problem only when building multiple modules.

great that we know where the problem is now!  I can try to look again next tues/weds when I'm back at my office (overseas at the moment, and my laptop doesn't have a scipy dev environment set up)

tupui commented:
Thank you 🙇 FYI to build the doc faster, I just trimmed down the file `API.rst.txt` to only build modules I wanted (and removed all corresponding files in `source/reference`), and I removed in `index.rst` the tutorial section which is also a large chunk as it runs a lot of code examples (I know it's ridiculous and we/I are trying to propose to move this section to another website/build phase).

choldgraf commented:
Could you clarify whether you think that the speed issue is related to this PR or this theme? If not then I think it would be better to open a dedicated issue to track down and discuss the autosummary issue.

tupui commented:
Sure, the issue is not linked to this PR. On main I have the same behaviour, when I have some autosummary template, it's slow. And if I remove them, build is almost twice as fast.

I can create a new issue if you want and we merge this?

drammock commented:
+1 for merge here.

choldgraf commented:
OK I haven't seen any objections and this one does come with a nice speed-up, so I'm gonna merge this :-)

drammock commented:
I cannot tell yet (haven't looked closely at their templates yet) but it's almost surely not related to the changes in *this PR*. Fine by me to open a new issue
-------- Original Message --------
On Aug 24, 2022, 05:30, Chris Holdgraf wrote:

> Could you clarify whether you think that the speed issue is related to this PR or this theme? If not then I think it would be better to open a dedicated issue to track down and discuss the autosummary issue.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/pydata/pydata-sphinx-theme/pull/878#issuecomment-1225176530), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAN2AUZYGVBDDYMT5EM5N7TV2WQMXANCNFSM56YJ4DMQ).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

jayaddison commented:
Connecting a reference that GitHub doesn't seem to have linked/displayed here, to make sure that doesn't get dropped - this could be related to a JSON encoding issue reported at sphinx-doc/sphinx#10990.
